subsys/logging/log_backend_cavs_hda.c

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "toolchain/gcc.h"
+#include <spinlock.h>
+#include <logging/log_backend.h>
+#include <logging/log_core.h>
+#include <logging/log_msg.h>
+#include <logging/log_output.h>
+#include <logging/log_backend_std.h>
+#include <logging/log_backend_cavs_hda.h>
+#include <drivers/dma.h>
+
+static uint32_t log_format_current = CONFIG_LOG_BACKEND_CAVS_HDA_OUTPUT_DEFAULT;
+static const struct device *hda_log_dev;
+static uint32_t hda_log_chan;
+
+/* Trace output goes in 64 byte chunks with a 4-byte header. HDA however
+ * requires 128 byte aligned data and 128 byte aligned transfers.
+ * So regardless of the log output we *always* send 128 bytes. The other
+ * side then can interpret it as needed.
+ */
+static __aligned(128) uint8_t hda_log_buf[CONFIG_LOG_BACKEND_CAVS_HDA_SIZE];
+static volatile uint32_t hda_log_buffered;
+static atomic_t hda_log_lock;
+static struct k_timer hda_log_timer;
+static cavs_hda_log_hook_t hook;
+
+static void hda_log_flush(void)
+{
+	/* Flushing is a best effort, if we fail now we can catch it a little
+	 * later
+	 */
+	if (!atomic_cas(&hda_log_lock, 0, 1)) {
+		return;
+	}
+
+	uint32_t nearest128 = hda_log_buffered & ~((128) - 1);
+
+	if (nearest128 > 0) {
+		hda_log_buffered = hda_log_buffered - nearest128;
+
+		dma_reload(hda_log_dev, hda_log_chan, 0, 0, nearest128);
+
+		if (hook != NULL) {
+			hook(nearest128);
+		}
+	}
+
+	atomic_set(&hda_log_lock, 0);
+}
+
+
+static int hda_log_out(uint8_t *data, size_t length, void *ctx)
+{
+	int ret;
+	struct dma_status status;
+
+	while (!atomic_cas(&hda_log_lock, 0, 1)) {
+	}
+
+	ret = dma_get_status(hda_log_dev, hda_log_chan, &status);
+	if (ret != 0) {
+		ret = 0;
+		goto exit;
+	}
+
+	/* The hardware tells us what space we have available, and where to
+	 * start writing. If the buffer is full we have no space.
+	 */
+	if (status.free <= 0) {
+		ret = 0;
```

Comment:
This callback is really weird, the doc says that it should return the number of bytes processed, dropped or discarded. Don't know how this should be done, if write 10 bytes out of 15 what should you return ? 10 or -5 ? Anyway, I think return 0 is fine, -ENOSPC also would be ok.

Reply:
yeah I wasn't sure either, 0 seems to work out ok, I'll look closer at it. 0 is probably wrong though

Reply:
Looking at how its used, I always need to return all the bytes here I believe since I'm not really excited to try and write some partial number of bytes. So I always return the full length of the message now.